src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/RdsHostUtils.java

```diff
@@ -0,0 +1,304 @@
+import java.util.regex.Pattern;
+
+import static com.mysql.cj.jdbc.ha.plugins.RdsUrlType.IP_ADDRESS;
+import static com.mysql.cj.jdbc.ha.plugins.RdsUrlType.OTHER;
+import static com.mysql.cj.jdbc.ha.plugins.RdsUrlType.RDS_CUSTOM_CLUSTER;
+import static com.mysql.cj.jdbc.ha.plugins.RdsUrlType.RDS_INSTANCE;
+import static com.mysql.cj.jdbc.ha.plugins.RdsUrlType.RDS_PROXY;
+import static com.mysql.cj.jdbc.ha.plugins.RdsUrlType.RDS_READER_CLUSTER;
+import static com.mysql.cj.jdbc.ha.plugins.RdsUrlType.RDS_WRITER_CLUSTER;
+
+public class RdsHostUtils {
+  public static final int NO_CONNECTION_INDEX = -1;
+  private final Pattern auroraDnsPattern =
+      Pattern.compile(
+          "(.+)\\.(proxy-|cluster-|cluster-ro-|cluster-custom-)?([a-zA-Z0-9]+\\.[a-zA-Z0-9\\-]+\\.rds\\.amazonaws\\.com)",
+          Pattern.CASE_INSENSITIVE);
+  private final Pattern rdsInstanceDnsPattern =
+      Pattern.compile(
+          "(.+)\\.(proxy-|cluster-|cluster-ro-|cluster-custom-){0}([a-zA-Z0-9]+\\.[a-zA-Z0-9\\-]+\\.rds\\.amazonaws\\.com)",
+          Pattern.CASE_INSENSITIVE);
+  private final Pattern auroraCustomClusterPattern =
+      Pattern.compile(
+          "(.+)\\.(cluster-custom-[a-zA-Z0-9]+\\.[a-zA-Z0-9\\-]+\\.rds\\.amazonaws\\.com)",
+          Pattern.CASE_INSENSITIVE);
+  private final Pattern auroraProxyDnsPattern =
+      Pattern.compile(
+          "(.+)\\.(proxy-[a-zA-Z0-9]+\\.[a-zA-Z0-9\\-]+\\.rds\\.amazonaws\\.com)",
+          Pattern.CASE_INSENSITIVE);
+  private final Log logger;
+
+  public RdsHostUtils(Log logger) {
+    this.logger = logger;
+  }
+
+  public RdsUrlType getUrlType(ConnectionUrl connectionUrl) {
+    return getUrlType(connectionUrl.getMainHost().getHost());
+  }
+
+  public RdsUrlType getUrlType(String url) {
+    if (IpAddressUtils.isIPv4(url) || IpAddressUtils.isIPv6(url)) {
+      return IP_ADDRESS;
+    } else if (isWriterClusterDns(url)) {
+      return RDS_WRITER_CLUSTER;
+    } else if (isReaderClusterDns(url)) {
+      return RDS_READER_CLUSTER;
+    } else if (isRdsCustomClusterDns(url)) {
+      return RDS_CUSTOM_CLUSTER;
+    } else if (isRdsProxyDns(url)) {
+      return RDS_PROXY;
+    } else if (isRdsDns(url)) {
+      return RDS_INSTANCE;
+    } else {
+      return OTHER;
+    }
+  }
+
+  public RdsHost getRdsHost(PropertySet propertySet, HostInfo hostInfo) throws SQLException {
+    final String hostName = hostInfo.getHost();
+    final int hostPort = hostInfo.getPort();
+    final RdsUrlType rdsUrlType = getUrlType(hostName);
+    final String clusterId = getClusterId(propertySet, rdsUrlType, hostName, hostPort);
+
+    String clusterInstanceTemplateHost = hostName;
+    if (rdsUrlType.isRds()) {
+      clusterInstanceTemplateHost = getRdsInstanceHostPattern(hostName);
+      if (clusterInstanceTemplateHost == null) {
+        this.logger.logError(Messages.getString("ConnectionStringTopologyProvider.5"));
+        throw new SQLException(Messages.getString("ConnectionStringTopologyProvider.5"));
```

Comment:
ConnectionStringTopologyProvider - probably need to change message id

Reply:
ah yeah my bad, thanks for catching that